[LoopVectorize] Add test case for minloc reduction

madhur13490 opened the pull request:
This patch adds a test case extracted from Polyhedron benchmark. https://fortran.uk/fortran-compiler-comparisons/the-polyhedron-solutions-benchmark-suite/

The test is specifically interesting for vectorizing min/max reduction pattern.

The actual FORTRAN code is

```
integer function minlst(ipos1, ipos2)
       integer, intent(in) :: ipos1, ipos2
       integer :: ipos
       minlst = ipos2
       do ipos = ipos2 - 1, ipos1, -1
           if (xxtrt(ipos) < xxtrt(minlst)) then
               minlst = ipos
           end if
       end do
   end function minlst
```
Patch to recognize and transform the pattern in LoopIdiomVectorize can be found at https://github.com/llvm/llvm-project/pull/144987

madhur13490 committed:
[LoopVectorize] Add test case for minloc reduction

This patch adds a test case extracted from Polyhedron benchmark.
https://fortran.uk/fortran-compiler-comparisons/the-polyhedron-solutions-benchmark-suite/

The test is specifically interesting for vectorizing min/max reduction
pattern.

madhur13490 commented:
Related PRs: https://github.com/llvm/llvm-project/pull/141431 and https://github.com/llvm/llvm-project/pull/141467.

fhahn commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,84 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -passes=loop-vectorize -S %s | FileCheck %s --check-prefix=CHECK-REV-MIN
```
This will need something like `-force-vector-width=4` to make sure this gets vectorized if possible.
```suggestion
; RUN: opt -passes=loop-vectorize -force-vector-width=4 -S %s | FileCheck %s
```

fhahn commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,84 @@
+  br label %.lr.ph
+
+.lr.ph:                                           ; preds = %.lr.ph.preheader, %.lr.ph
+  %indvars.iv = phi i64 [ %7, %.lr.ph.preheader ], [ %indvars.iv.next, %.lr.ph ]
```
```suggestion
  %iv = phi i64 [ %7, %.lr.ph.preheader ], [ %indvars.iv.next, %.lr.ph ]
```

fhahn reviewed (commented):
Thamks for adding the test. Left some initial comments

madhur13490 committed:
Address review comments

madhur13490 commented:
@fhahn Thanks for the review. I have addressed comments, I hope that covers all!

madhur13490 commented:
@fhahn or @Mel-Chen Would any of your pending patches cause vectorization of this test case? (I am having LoopIdiomVectorize based patch ready, and I should be upstreaming that soon, but wanted to check once.)

Mel-Chen commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,86 @@
+  br label %loop
+
+loop:                                           ; preds = %loop.preheader, %loop
+  %iv = phi i64 [ %last_index_sext, %loop.preheader ], [ %iv.next, %loop ]
+  %dec_iv = phi i64 [ %dec, %loop ], [ %diff, %loop.preheader ]
```
Do you really need two induction variable with same step?

madhur13490 replied:
Yes.

Mel-Chen commented:
> Related PRs: #141431 and #141467.

No, I think this is https://github.com/llvm/llvm-project/pull/140451.
There is only one phi node involved the reduction.

madhur13490 committed:
Address review comments

madhur13490 commented:
Hi @Mel-Chen Thanks for the review. I have addressed some of your comments. For others, I want to preserve the structure of the code generated by the Flang compiler. As said, this is extracted from Flang pipeline and particularly the IR is just before `LoopIdiomVectorize` pass, thus we have LCSSA structure preserved.

madhur13490 commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,86 @@
+  %last_index_neg_sext = sext i32 %last_index_neg to i64
+  %add = add nsw i64 %first_index_sext, %last_index_neg_sext
+  %diff = sub nsw i64 0, %add
+  %first_ptr = getelementptr i8, ptr %array, i64 -8
+  %second_ptr = getelementptr i8, ptr %array, i64 -4
```
No, I'd like to preserve the structure.

Mel-Chen replied:
Why? Would passing them directly prevent this case from being vectorized?

fhahn replied:
The main part of the test is to check we can vectorize this particular idiom, which should be independent of anything frontend-specific.

You could add a separate more complex test in addition, but I don't see how `%early_exit_cond and the sign extends are relevant, IIUC they could just be passed as wider types.

madhur13490 commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,86 @@
+; The function minlst primarily takes two indices (i.e. range), scans backwards in the range
+; and returns the firstIV of the minimum value.
+
+define fastcc i32 @minlst(i32 %first_index, i32 %last_index, ptr %array) {
```
Yes.

madhur13490 commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,86 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -passes=loop-vectorize -force-vector-width=4 -S %s | FileCheck %s --check-prefix=CHECK-REV-MIN
```
Done.

Mel-Chen commented:
> Hi @Mel-Chen Thanks for the review. I have addressed some of your comments. For others, I want to preserve the structure of the code generated by the Flang compiler. As said, this is extracted from Flang pipeline and particularly the IR is just before `LoopIdiomVectorize` pass, thus we have LCSSA structure preserved.

Generally, we hope test cases to be as precise and clean as possible. So unless the input IR not being in LCSSA form is the reason LoopVectorize or LoopIdiomVectorize doesn't work, it's better to clean up the test case.

madhur13490 committed:
Address review comments

madhur13490 commented:
Gentle ping @Mel-Chen. I have made the changes except one on the passing pointers as args. Please have a look.

madhur13490 committed:
Address review comments and add complex version

madhur13490 commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,198 @@
+  br label %loop
+
+loop:                                           ; preds = %loop.preheader, %loop
+  %iv = phi i64 [%iv.next, %loop], [ %last_index_sext, %loop.preheader ]
```
Done.

Mel-Chen commented on llvm/test/Transforms/LoopVectorize/last-min-index.ll:
```diff
@@ -0,0 +1,198 @@
+  %load1 = load float, ptr %load1_ptr, align 4
+  %index_sext = sext i32 %index to i64
+  %load2_ptr = getelementptr float, ptr %second_ptr, i64 %index_sext
+  %load2 = load float, ptr %load2_ptr, align 4
```
I think there might have been a misunderstanding. What I meant was whether this case could potentially be vectorized in the future. The reason I brought it up is: 
>  The value of %load2 depends on the result of %index from the previous iteration. Would this dependency prevent %load2 from being directly widened into a vector load?

This goes beyond just recurrence analysis.
By the way, is this intended as a pre-commit test case? If so, is there already a corresponding patch for it?

david-arm replied:
OK, but PR https://github.com/llvm/llvm-project/pull/144987 is not a LoopVectorize change.

madhur13490 replied:
This case demonstrates where LV fails to perform vectorization; thus, I added it in LoopVectorize. I don't mind it moving to LoopIdiomVectorize, though, if we prefer.

david-arm commented on llvm/test/Transforms/LoopVectorize/last-min-index-ftn.ll:
```diff
@@ -0,0 +1,160 @@
+  %load1 = load float, ptr %load1_ptr, align 4
+  %index_sext = sext i32 %index to i64
+  %load2_ptr = getelementptr float, ptr %second_ptr, i64 %index_sext
+  %load2 = load float, ptr %load2_ptr, align 4
```
I agree with @Mel-Chen, this looks very difficult to loop-vectorise because the pointer depends upon a value loaded from the previous iteration. Surely, the root problem here is that LoadPRE in an earlier IR pass has failed to spot this recurrence and keep a copy of the last loaded value corresponding to `array[%index]` in a phi node? For example, if the IR looked more like this:

```
loop:                                           ; preds = %loop.preheader, %loop
  %iv = phi i64 [%iv.next, %loop], [ %last_index_sext, %loop.preheader ]
  %dec_iv = phi i64 [ %dec, %loop ], [ %diff, %loop.preheader ]
  %index = phi i32 [ %select, %loop ], [ %last_index, %loop.preheader ]
  %val_at_index = phi float [ %new_val_at_index, %loop ], [ ... ]
  %iv.next = add nsw i64 %iv, -1
  %load1_ptr = getelementptr float, ptr %first_ptr, i64 %iv
  %load1 = load float, ptr %load1_ptr, align 4
  %cmp = fcmp contract olt float %load1, %val_at_index
  %iv.next.trunc = trunc nsw i64 %iv.next to i32
  %select = select i1 %cmp, i32 %iv.next.trunc, i32 %index
  %val_at_index = select i1 %cmp, %load1, %val_at_index
  %dec = add nsw i64 %dec_iv, -1
  %loop_cond = icmp sgt i64 %dec_iv, 1
  br i1 %loop_cond, label %loop, label %._crit_edge
```

not only would the scalar version be faster, but it might give the loop vectoriser a better chance of using existing min/max reduction infrastructure to vectorise it?

david-arm replied:
Perhaps I'm missing some critical limitation here, but it feels like the focus we should be trying to optimise the scalar IR before we attempt vectorisation. There are no stores in the loop and the bounds seem to be well-known.

madhur13490 replied:
Currently the limitation in GVN (if we go by what GCC does). GVN is GCC is much more sophisticated than what LLVM has and improving LLVM's GVN is a large effort. However, having said that, optimizing scalar code would lead to 10-12% gain - as we would eliminate just the PRE case, vectorization leads to 4x speedup on an average.

fhahn replied:
Right, but when handled in LoopIdiom, it would also need to perform this transform at least implicitly. So it should also be possible to perform the PRE independently

madhur13490 commented on llvm/test/Transforms/LoopVectorize/last-min-index-ftn.ll:
```diff
@@ -0,0 +1,160 @@
+  %iv = phi i64 [%iv.next, %loop], [ %last_index_sext, %loop.preheader ]
+  %dec_iv = phi i64 [ %dec, %loop ], [ %diff, %loop.preheader ]
+  %index = phi i32 [ %select, %loop ], [ %last_index, %loop.preheader ]
+  %iv.next = add nsw i64 %iv, -1
```
Please see Fortran source code. The source has reverse-counting loop thus we it in IR; unfortunately we can't generalize this.

madhur13490 replied:
Ok, I moved the test *-ftn to LoopIdiom and keeping the other one in LV.

madhur13490 committed:
Move fortran test to LoopIdiom

madhur13490 commented on llvm/test/Transforms/LoopVectorize/last-min-index-ftn.ll:
```diff
@@ -0,0 +1,160 @@
+; RUN: opt -passes=loop-vectorize -force-vector-width=4 -force-vector-interleave=1 -S %s | FileCheck %s --check-prefix=CHECK-REV-MIN-VW4-IL1
+; RUN: opt -passes=loop-vectorize -force-vector-width=4 -force-vector-interleave=2 -S %s | FileCheck %s --check-prefix=CHECK-REV-MIN-VW4-IL2
+
+; This test case is extracted from rnflow (fortran) benchmark in polyhedron benchmark suite.
```
This is very specific to one benchmark because of how the loop is written. You can find the actual fortran loop in patch description.
